skcosmo/preprocessing/flexible_scaler.py

```diff
@@ -212,9 +213,108 @@ def fit_transform(self, K, y=None, ref_cmean=None, ref_mean=None, **fit_params):
         :type ref_cmean: array of shape (1, n_features)
         :param ref_mean: an average for the whole kernel matrix
         :type ref_mean: array
-        :param \**fit_params: necessary for compatibility with the functions of the TransformerMixin class
+        :param \**fit_params: necessary for compatibility with the functions of
+                              the TransformerMixin class
 
         :return: tranformed matrix Kc
         """
         self.fit(K, y, ref_cmean=ref_cmean, ref_mean=ref_mean)
         return self.transform(K, y)
+
+
+class SparseKernelCenterer(TransformerMixin, BaseEstimator):
+    """Kernel centering method for sparse kernels, similar to
+    KernelFlexibleCenterer
+    """
+
+    def __init__(self, rcond=1e-12):
+        """
+        Initialize SparseKernelCenterer.
+
+        :param rcond: conditioning parameter to use when computing the
+                      Nystrom-approximated kernel for scaling
+        :type rcond: float, default 1E-12
+        """
+
+        self.rcond = rcond
+
+    def fit(self, Knm, Kmm, y=None):
+        """Fit KernelFlexibleCenterer
+
+        :param Knm: Kernel matrix between the reference data set and the active
+                    set
+        :type Knm: ndarray of shape (n_samples, n_active)
+
+        :param Kmm: Kernel matrix between the active set and itself
+        :type Kmm: ndarray of shape (n_active, n_active)
+
+        :param y: ignored
+
+        :return: itself
+        """
+
+        if Knm.shape[1] != Kmm.shape[0]:
+            raise ValueError(
+                "The reference kernel is not commensurate shape with the"
+                "active kernel."
+            )
+
+        if Kmm.shape[0] != Kmm.shape[1]:
+            raise ValueError("The active kernel is not square.")
+
+        self.n_active_ = Kmm.shape[0]
+
+        self.K_fit_rows_ = Knm.mean(axis=0)
+
+        Knm_centered = Knm - np.broadcast_arrays(Knm, self.K_fit_rows_)[1]
+        Kmm_centered = Kmm - np.broadcast_arrays(Kmm, self.K_fit_rows_)[1]
```

Comment:
Centering `Kmm` is a bit of a tricky subject. Long story short, I would probably leave it uncentered. In all my tests of this, it doesn't matter if you don't center `Kmm`, center it by itself (using a "traditional" square kernel centering) or do the subtraction like you have here -- the resulting scaled/centered `Knm` and Nystrom kernel approximation are all the same, and have the appropriate row and column means and trace. You might want to check this on your own just to be sure. Writing out the sparse centering analytically, there isn't any requirement to center `Kmm` and so I think it might be best (or at least more consistent) to leave it uncentered.

Reply:
I removed this centering, as I ran into similar issues